rsocket/RSocketClient.cpp

```diff
@@ -35,28 +35,44 @@ folly::Future<std::unique_ptr<RSocketRequester>> RSocketClient::connect(
   folly::Promise<std::unique_ptr<RSocketRequester>> promise;
   auto future = promise.getFuture();
 
-  connectionFactory_->connect([
-    this,
-    setupParameters = std::move(setupParameters),
-    responder = std::move(responder),
-    keepaliveTimer = std::move(keepaliveTimer),
-    stats = std::move(stats),
-    networkStats = std::move(networkStats),
-    promise = std::move(promise)](
-      std::unique_ptr<DuplexConnection> connection,
-      folly::EventBase& eventBase) mutable {
-    VLOG(3) << "onConnect received DuplexConnection";
-
-    auto rsocket = fromConnection(
-        std::move(connection),
-        eventBase,
-        std::move(setupParameters),
-        std::move(responder),
-        std::move(keepaliveTimer),
-        std::move(stats),
-        std::move(networkStats));
-    promise.setValue(std::move(rsocket));
-  });
+  connectionFactory_->connect()
+      .then([
+        this,
+        setupParameters = std::move(setupParameters),
+        responder = std::move(responder),
+        keepaliveTimer = std::move(keepaliveTimer),
+        stats = std::move(stats),
+        networkStats = std::move(networkStats),
+        promise = std::move(promise)
+      ](ConnectionResult connResult) mutable {
+        VLOG(3) << "onConnect received DuplexConnection";
+        // Ensure we create the RSocketStateMachine in the right evb.
+        connResult.evb.runImmediatelyOrRunInEventBaseThreadAndWait([
```

Comment:
Instead of this you should be able to ensure the `then()` clause happens on the correct eventbase using `via()` on the caller side.